rust/core-lib/src/movement.rs

```diff
@@ -80,20 +84,48 @@ fn vertical_motion(r: SelRegion, view: &View, text: &Rope, line_delta: isize,
     // This code is quite careful to avoid integer overflow.
     // TODO: write tests to verify
     let line = view.line_of_offset(text, active);
+    let mut line_length = view.offset_of_line(text, line.saturating_add(1)) - view.offset_of_line(text, line);
     if line_delta < 0 && (-line_delta as usize) > line {
-        return (0, Some(col));
+        if enforce_horiz_pos {
```

Comment:
thinking out loud: you could write this bit,

```rust
let row = if enforce_horiz_pos { active } else { 0 };
return (row, Some(col));
```

which _might_ be an improvement? Not a huge win though.

Reply:
If I don't split the methods I will make this change.

Reply:
Yea my main motivation is just how many different exit points this function has; it gets quite hard to follow. Another option might be to do all the setup in a shared function, and then branch afterwards? Just thinking out loud.